mypy.ini

```diff
@@ -9,83 +9,189 @@ mypy_path = stubs
 warn_unreachable = True
 local_partial_types = True
 no_implicit_optional = True
+scripts_are_modules = True
 
-# To find all folders that pass mypy you run:
-#
-#   find synapse/* -type d -not -name __pycache__ -exec bash -c "mypy '{}' > /dev/null"  \; -print
+# To find files which are Python scripts, but which do not end in .py, run:
+#   rg --glob '!*.py' '^#!.*python' | cut -d ':' -f 1 | sort --unique
 
 files =
-  scripts-dev/sign_json,
-  synapse/__init__.py,
-  synapse/api,
-  synapse/appservice,
-  synapse/config,
-  synapse/crypto,
-  synapse/event_auth.py,
-  synapse/events,
-  synapse/federation,
-  synapse/groups,
-  synapse/handlers,
-  synapse/http,
-  synapse/logging,
-  synapse/metrics,
-  synapse/module_api,
-  synapse/notifier.py,
-  synapse/push,
-  synapse/replication,
-  synapse/rest,
-  synapse/server.py,
-  synapse/server_notices,
-  synapse/spam_checker_api,
-  synapse/state,
-  synapse/storage/__init__.py,
-  synapse/storage/_base.py,
-  synapse/storage/background_updates.py,
-  synapse/storage/databases/main/appservice.py,
-  synapse/storage/databases/main/client_ips.py,
-  synapse/storage/databases/main/events.py,
-  synapse/storage/databases/main/keys.py,
-  synapse/storage/databases/main/pusher.py,
-  synapse/storage/databases/main/registration.py,
-  synapse/storage/databases/main/relations.py,
-  synapse/storage/databases/main/session.py,
-  synapse/storage/databases/main/stream.py,
-  synapse/storage/databases/main/ui_auth.py,
-  synapse/storage/databases/state,
-  synapse/storage/database.py,
-  synapse/storage/engines,
-  synapse/storage/keys.py,
-  synapse/storage/persist_events.py,
-  synapse/storage/prepare_database.py,
-  synapse/storage/purge_events.py,
-  synapse/storage/push_rule.py,
-  synapse/storage/relations.py,
-  synapse/storage/roommember.py,
-  synapse/storage/state.py,
-  synapse/storage/types.py,
-  synapse/storage/util,
-  synapse/streams,
-  synapse/types.py,
-  synapse/util,
-  synapse/visibility.py,
-  tests/replication,
-  tests/test_event_auth.py,
-  tests/test_utils,
-  tests/handlers/test_password_providers.py,
-  tests/handlers/test_room.py,
-  tests/handlers/test_room_summary.py,
-  tests/handlers/test_send_email.py,
-  tests/handlers/test_sync.py,
-  tests/handlers/test_user_directory.py,
-  tests/rest/client/test_login.py,
-  tests/rest/client/test_auth.py,
-  tests/rest/client/test_relations.py,
-  tests/rest/media/v1/test_filepath.py,
-  tests/rest/media/v1/test_oembed.py,
-  tests/storage/test_state.py,
-  tests/storage/test_user_directory.py,
-  tests/util/test_itertools.py,
-  tests/util/test_stream_change_cache.py
+  # Discover all .py[i] files in the repo
+  .,
```

Comment:
won't this iterate into virtualenvs etc?

Reply:
Mypy won't recurse into hidden directories, so as long as your venv is in `.venv/` you're golden.

Are there any other common paths we should specifically exclude?

Reply:
The README (and maybe docs) recommends putting the venv in an `env` directory with `python3 -m venv ./env`. I did this on day one.

Reply:
Hm. Y'all should definitely switch to `.venv`, but I'll add an exclusion for `env/` for now ;)

Reply:
Happy to; I was just getting at that we ought to update the README! (and perhaps for sydent/sygnal too?)

Reply:
I'm not enthusiastic about adding this sort of constraint to people's dev environments. I'm not aware of any rule that says "thou shalt have no virtualenvs in your working tree save that they be called `.venv`", and I don't necessarily expect anyone else to be aware of such a rule.

Please can we just list the directories we care about?

Reply:
To add to this: my main concern is if every different project starts having its own, specific rules about where the venv should go. I really don't want to have to start remembering different places for different projects.

So, if this is a thing we want to do, we have to get into a big debate about what The One True Place For Your Virtualenv should be, which isn't a fight I particularly want to have.

Reply:
also also: I have several *different* venvs in my working copy, with different selections of libraries, and python interpreters, so in any case just excluding `env` doesn't help a great deal.

Reply:
/me gestures wildly at "I'll add an exclusion for env/ for now ;)" and 515ca6bb6c085cd94d1ccf11bd07faf81ea95fde from an hour ago

Reply:
I'm yet to be convinced, I'm afraid.

Reply:
basically anything that's not checked in - my whole point is that this shouldn't break when people have things in their working copy that you don't expect.

Reply:
I guess maybe it would be ok to ignore things that are in the user's various gitignore configs, but that feels a bit magical.

Reply:
In [Apache voting terms](http://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions), how strongly are your feelings on this issue?

Assume the alternative would be to only cover `synapse/` and `tests/` with mypy, allowing other files in the repo to go without type checking.

Reply:
> In [Apache voting terms](http://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions), how strongly are your feelings on this issue?

-0.9ish.

I like this way of resolving the dispute.

> Assume the alternative would be to only cover synapse/ and tests/ with mypy, allowing other files in the repo to go without type checking.

Well, we could surely add other directories - `scripts-dev`, `scripts`, etc. As noted elsewhere, we already have to have multiple entries in the list.